File: Easy/4- 53. Maximum Subarray/53. Maximum Subarray.cpp
// Tutorial: https://backtobackswe.com/platform/content/max-contiguous-subarray-sum/solutions

// Important link: https://www.geeksforgeeks.org/climits-limits-h-cc/
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int maxSoFar=nums[0];
        int maxEndingHere=nums[0];
        
        for(int i=1;i<nums.size();i++)
        {
            maxEndingHere=max(maxEndingHere+nums[i],nums[i]);
            maxSoFar=max(maxSoFar,maxEndingHere);
        }
        
        return maxSoFar;
    }
};

File: Easy/4- 53. Maximum Subarray/53. Maximum Subarray.cs
//https://backtobackswe.com/platform/content/max-contiguous-subarray-sum/solutions
public class Solution {
    public int MaxSubArray(int[] nums) {
            int maxSoFar = nums[0];
            int maxEndingHere = nums[0];

            for (int i = 1; i < nums.Length; i++)
            {

                //Console.WriteLine($"maxEndingHere: {maxEndingHere}");
                //Console.WriteLine($"max of ({maxEndingHere+nums[i]}, {nums[i]})");
                maxEndingHere = Math.Max(maxEndingHere + nums[i], nums[i]);
                //Console.WriteLine($"------->maxEndingHere: {maxEndingHere}");
                //Console.WriteLine($"maxSoFar: {maxSoFar}");
                maxSoFar = Math.Max(maxSoFar, maxEndingHere);
                //Console.WriteLine($"------->Final maxSoFar: {maxSoFar}");
                //Console.WriteLine();
            }

            return maxSoFar;
        }
}
